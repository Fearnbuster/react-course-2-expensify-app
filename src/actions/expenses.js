
import firestore from '../firebase/firebase';

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return firestore.collection('expenses')
      .add(expense)
      .then((newExpense) => {
        dispatch(addExpense({
          id: newExpense.id,
          ...expense
        }));
      });
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

