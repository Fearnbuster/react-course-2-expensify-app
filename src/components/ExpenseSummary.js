
import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

const createSummaryMessage = (expenses) => {
  const numberOfExpenses = expenses.length;
  const totalExpense = numeral(getExpensesTotal(expenses) / 100).format('$0,0.00');

  switch (numberOfExpenses) {
    case 0:
      return '';

    case 1:
      return <p>
          {`Viewing 1 expense totalling ${totalExpense}`}
        </p>
  
    default:
      return <p>
        {`Viewing ${numberOfExpenses} expenses totalling ${totalExpense}`}
      </p>
  }
};

export const ExpenseSummary = (props) => (
  <div>
    {createSummaryMessage(props.expenses)}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseSummary);