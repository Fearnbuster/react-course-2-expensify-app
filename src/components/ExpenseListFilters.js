
import React from 'react';
import { connect } from 'react-redux';

import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

import selectExpenses from '../selectors/expenses';
import { start } from 'repl';

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    switch (e.target.value) {
      case 'amount': {
        this.props.sortByAmount();
        break;
      }

      case 'date': {
        this.props.sortByDate();
        break;
      }
    }
  };

  render() {
    return (
      <div>
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
        <select defaultValue="date" onChange={this.onSortChange}>
          <option value="amount">Amount</option>
          <option value="date">Date</option>
        </select>
        <DateRangePicker
          startDateId="startDatePicker"
          endDateId="endDatePicker"
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters),
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);