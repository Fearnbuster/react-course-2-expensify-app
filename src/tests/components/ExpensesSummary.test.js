
import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

describe('components/ExpensesSummary', () => {
  it('should render component correctly with 1 expense', () => {
    const wrapper = shallow(
      <ExpenseSummary 
        expenses={[expenses[0]]}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render component correctly with multiple expenses', () => {
    const wrapper = shallow(
      <ExpenseSummary 
        expenses={expenses}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});