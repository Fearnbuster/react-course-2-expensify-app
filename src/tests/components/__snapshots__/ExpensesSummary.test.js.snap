// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ExpensesSummary should render component correctly with 1 expense 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $1.95
  </p>
</div>
`;

exports[`components/ExpensesSummary should render component correctly with multiple expenses 1`] = `
<div>
  <p>
    Viewing 3 expenses totalling $1,141.95
  </p>
</div>
`;
